Add derivation for firefox-bin on macOS

Enzime opened the pull request:
###### Description of changes

Fixes #71689

Had to rewrite some of `maintainers/scripts/update.py` as I was getting deadlocks due to there being too much `stderr` output when running it with `firefox-bin/update.nix`

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Enzime committed:
firefox-bin: add Mac sources

Enzime committed:
firefox-bin: Add Mac arch to updateScript

jtojnar reviewed (commented):
Nice catch, looks like the deadlock is mentioned twice in the docs:

- https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.subprocess.Process.wait
- https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.subprocess.Process.stderr

Could you please make the `update.py` into a separate commit?

Enzime committed:
maintainers/scripts/update.nix: avoid deadlock

Use `Process.communicate()` instead of `Process.wait()` to ensure the
`stdin` and `stdout` OS pipe buffers don't get full and cause a deadlock
waiting for the buffers to get emptied.

Enzime commented on maintainers/scripts/update.py:
```diff
@@ -21,12 +21,14 @@ def eprint(*args, **kwargs):
-    returncode = await process.wait()
+    # We need to use communicate() instead of wait(), as the OS pipe buffers
+    # can fill up and cause a deadlock.
+    process.stdout, process.stderr = await process.communicate()
```
Updated :+1:

Enzime commented:
@jtojnar separated `update.py` into another commit :)

jtojnar commented on maintainers/scripts/update.py:
```diff
@@ -12,27 +12,31 @@
     print(*args, file=sys.stderr, **kwargs)
 
-async def check_subprocess(*args, **kwargs):
+async def check_subprocess_output(*args, **kwargs):
```
I would use `outputs` or `output_and_error` to distinguish it from `subprocess.check_output`, which might give people reading the code an incorrect expectation.

Enzime replied:
Updated to `check_subprocess_outputs` :+1:

Enzime commented:
@jtojnar updated the function name :)

utkarshgupta137 commented:
Any updates?

Atemu commented:
@jtojnar could you review the update script part again?

Atemu commented:
@Enzime you need to rebase.

jtojnar commented on maintainers/scripts/update.py:
```diff
@@ -12,27 +12,31 @@
 
 class CalledProcessError(Exception):
     process: asyncio.subprocess.Process
+    stderr: Optional[bytes]
```
Is there a case where this can be `None`? It seems to me we are only setting it in the one place?

jtojnar commented on maintainers/scripts/update.py:
```diff
@@ -127,12 +130,12 @@ async def check_changes(package: Dict, worktree: str, update_info: str):
             attr_path = changes[0]['attrPath']
-            obtain_new_version_process = await check_subprocess('nix-instantiate', '--expr', f'with import ./. {{}}; lib.getVersion {attr_path}', '--eval', '--strict', '--json', stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE, cwd=worktree)
-            changes[0]['newVersion'] = json.loads((await obtain_new_version_process.stdout.read()).decode('utf-8'))
+            obtain_new_version_output = await check_subprocess_outputs('nix-instantiate', '--expr', f'with import ./. {{}}; lib.getVersion {attr_path}', '--eval', '--strict', '--json', stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE, cwd=worktree)
```
Forgot to destruct the tuple here:

```suggestion
            obtain_new_version_output, _ = await check_subprocess_outputs('nix-instantiate', '--expr', f'with import ./. {{}}; lib.getVersion {attr_path}', '--eval', '--strict', '--json', stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE, cwd=worktree)
```

Enzime replied:
Fixed :+1:

Enzime replied:
Fixed :+1:

Enzime commented:
@jtojnar @Atemu updated my PR, can you guys take a look?

Enzime commented:
@jtojnar updated, renamed `check_subprocess_outputs` to `check_subprocess_output` now that it no longer returns `stderr` :+1:

Enzime commented:
@jtojnar can you take a look at this PR when you get a chance?

hraban commented:
Fwiw this doesn't work for me on darwin-aarch64:

```
$ nix build  github:Enzime/nixpkgs/firefox-bin-darwin#firefox-bin --show-trace
error:
       … while calling the 'derivationStrict' builtin

         at //builtin/derivation.nix:9:12: (source not available)

       … while evaluating derivation 'firefox-bin-112.0'
         whose name attribute is located at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/pkgs/stdenv/generic/make-derivation.nix:302:7

       … while evaluating attribute 'buildInputs' of derivation 'firefox-bin-112.0'

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/pkgs/stdenv/generic/make-derivation.nix:349:7:

          348|       depsHostHost                = lib.elemAt (lib.elemAt dependencies 1) 0;
          349|       buildInputs                 = lib.elemAt (lib.elemAt dependencies 1) 1;
             |       ^
          350|       depsTargetTarget            = lib.elemAt (lib.elemAt dependencies 2) 0;

       … while calling 'getOutput'

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/lib/attrsets.nix:892:23:

          891|   */
          892|   getOutput = output: pkg:
             |                       ^
          893|     if ! pkg ? outputSpecified || ! pkg.outputSpecified

       … while calling anonymous lambda

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/pkgs/stdenv/generic/make-derivation.nix:249:13:

          248|       (map (drv: drv.__spliced.hostHost or drv) (checkDependencyList "depsHostHost" depsHostHost))
          249|       (map (drv: drv.__spliced.hostTarget or drv) (checkDependencyList "buildInputs" buildInputs'))
             |             ^
          250|     ]

       … while calling anonymous lambda

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/lib/lists.nix:117:29:

          116|   */
          117|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                             ^
          118|

       … from call site

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/lib/lists.nix:117:32:

          116|   */
          117|   imap1 = f: list: genList (n: f (n + 1) (elemAt list n)) (length list);
             |                                ^
          118|

       … while calling anonymous lambda

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/pkgs/stdenv/generic/make-derivation.nix:216:81:

          215|   checkDependencyList = checkDependencyList' [];
          216|   checkDependencyList' = positions: name: deps: lib.flip lib.imap1 deps (index: dep:
             |                                                                                 ^
          217|     if lib.isDerivation dep || dep == null || builtins.typeOf dep == "string" || builtins.typeOf dep == "path" then dep

       … from call site

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/pkgs/stdenv/generic/make-derivation.nix:217:8:

          216|   checkDependencyList' = positions: name: deps: lib.flip lib.imap1 deps (index: dep:
          217|     if lib.isDerivation dep || dep == null || builtins.typeOf dep == "string" || builtins.typeOf dep == "path" then dep
             |        ^
          218|     else if lib.isList dep then checkDependencyList' ([index] ++ positions) name dep

       … while calling 'isDerivation'

         at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/lib/attrsets.nix:636:5:

          635|     # Value to check.
          636|     value: value.type or null == "derivation";
             |     ^
          637|

       error: attribute 'gtk3' missing

       at /nix/store/cw47zmdcab6hd787ws7arb87j482v356-source/pkgs/applications/networking/browsers/firefox/wrapper.nix:231:23:

          230|       nativeBuildInputs = [ makeWrapper lndir jq ];
          231|       buildInputs = [ browser.gtk3 ];
             |                       ^
          232|
```

Enzime commented:
@hraban the Firefox wrapper is currently Linux only, so the package you want is `firefox-bin-unwrapped`

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox-bin/update.nix:
```diff
@@ -56,29 +56,36 @@ in writeScript "update-${pname}" ''
   # this is a list of sha256 and tarballs for both arches
   # Upstream files contains python repr strings like b'somehash', hence the sed dance
-  shasums=`cat $HOME/shasums | sed -E s/"b'([a-f0-9]{64})'?(.*)"/'\1\2'/ | grep tar.bz2`
+  shasums=`cat $HOME/shasums | sed -E s/"b'([a-f0-9]{64})'?(.*)"/'\1\2'/ | grep "\(tar.bz2\|dmg\)" | grep -v mac-EME-free`
```
```suggestion
  shasums=$(cat $HOME/shasums | sed -E s/"b'([a-f0-9]{64})'?(.*)"/'\1\2'/ | grep "\(tar.bz2\|dmg\)" | grep -v mac-EME-free)
```

Enzime replied:
Fixed :+1:

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox-bin/default.nix:
```diff
@@ -214,4 +103,138 @@ stdenv.mkDerivation {
+      cp -r *.app $out/Applications
+    '';
+
+    passthru.updateScript = updateScript;
```
```suggestion
    passthru = {
      inherit updateScript;
    };
```

Enzime replied:
Fixed :+1:

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox-bin/default.nix:
```diff
@@ -214,4 +103,138 @@ stdenv.mkDerivation {
+        ln -s ${policiesJson} "$out/lib/firefox-bin-${version}/distribution/policies.json";
+      '';
+
+    passthru.binaryName = "firefox";
+    passthru.libName = "firefox-bin-${version}";
+    passthru.execdir = "/bin";
+    passthru.ffmpegSupport = true;
+    passthru.gssSupport = true;
+    passthru.updateScript = updateScript;
```
```suggestion
    passthru = {
      binaryName = "firefox";
      libName = "firefox-bin-${version}";
      execdir = "/bin";
      ffmpegSupport = true;
      gssSupport = true;
      inherit updateScript;
    };
```

Enzime replied:
Fixed :+1:

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox-bin/default.nix:
```diff
@@ -214,4 +103,138 @@ stdenv.mkDerivation {
+      echo 'pref("app.update.auto", "false");' >> defaults/pref/channel-prefs.js
+    '';
+
+    installPhase =
+      ''
```
```suggestion
    installPhase = ''
```

Enzime replied:
Fixed :+1:

SuperSandro2000 commented on pkgs/applications/networking/browsers/firefox-bin/default.nix:
```diff
@@ -214,4 +103,138 @@ stdenv.mkDerivation {
+  linux = stdenv.mkDerivation {
+    inherit pname version src meta;
+
+    libPath = lib.makeLibraryPath
+      [ stdenv.cc.cc
```
```suggestion
    libPath = lib.makeLibraryPath [
      stdenv.cc.cc
```

Enzime replied:
Fixed :+1:

tjni reviewed (commented):
The code LGTM. This sort of feels like a place where the derivations diverge so much that we'd want to keep the Linux one in this file as the main focus and extract the Darwin one into its own neighboring file.

Atemu commented:
> This sort of feels like a place where the derivations diverge so much that we'd want to keep the Linux one in this file as the main focus and extract the Darwin one into its own neighboring file.

I'd be in favour of that aswell. See https://github.com/NixOS/nixpkgs/pull/195473

hraban commented:
> @hraban the Firefox wrapper is currently Linux only, so the package you want is `firefox-bin-unwrapped`

thanks, can confirm this works 👍

Enzime commented:
@tjni @Atemu I intentionally chose to keep the files together as `firefox-bin` on macOS should stay quite simple as it'll just be a derivation that extracts a DMG and to avoid #227449

Atemu commented:
@Enzime that bug happens due to they way they chose to go about it. It'd be better to do that in all-packages IMO i.e. `if stdenv.isDarwin then callPackage ../path/to/spotify/darwin.nix { } else callPackage ../path/to/spotify { };`

lafrenierejm commented on pkgs/applications/networking/browsers/firefox-bin/update.nix:
```diff
@@ -56,29 +56,36 @@ in writeScript "update-${pname}" ''
-      { url = "$url$version/`echo $line | cut -d":" -f3`";
-        locale = "`echo $line | cut -d":" -f3 | sed "s/$arch\///" | sed "s/\/.*//"`";
+      { url = "$url$version/`echo "$line" | cut -d":" -f3`";
+        locale = "`echo "$line" | cut -d":" -f3 | sed "s/$arch\///" | sed "s/\/.*//"`";
```
```suggestion
        locale = "$(echo "$line" | cut -d":" -f3 | sed "s/$arch\///" | sed "s/\/.*//")";
```

Enzime replied:
Fixed :+1: \

Enzime committed:
firefox-bin/update.nix: fix `shellcheck` issues

Enzime committed:
firefox-bin: support macOS

Enzime commented:
I ran `shellcheck` on the Bash script in `update.nix` and fixed all the issues, and I separated `default.nix` into `darwin.nix` and `linux.nix`

Enzime committed:
firefox-bin: improve formatting

Atemu commented on pkgs/applications/networking/browsers/firefox-bin/darwin.nix:
```diff
@@ -0,0 +1,77 @@
+    license = licenses.mpl20;
+    sourceProvenance = with sourceTypes; [ binaryNativeCode ];
+    platforms = [ "x86_64-darwin" "aarch64-darwin" ];
+    hydraPlatforms = [ ];
```
Why not actually? It's a tarball-ish download either way.

Enzime replied:
Basing it off https://github.com/NixOS/nixpkgs/commit/b1a7f33b3927eede727144d1761c0afe61676706

Atemu commented on pkgs/applications/networking/browsers/firefox-bin/darwin.nix:
```diff
@@ -0,0 +1,77 @@
+, fetchurl
+, config
+, curl
+, adwaita-icon-theme
```
Unused input. Please clean these up.

Enzime replied:
Fixed :+1:

Atemu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -31373,7 +31373,7 @@ with pkgs;
   firefox-esr-115 = wrapFirefox firefox-esr-115-unwrapped { };
 
-  firefox-bin-unwrapped = callPackage ../applications/networking/browsers/firefox-bin {
+  firefox-bin-unwrapped = callPackage (if stdenv.isDarwin then ../applications/networking/browsers/firefox-bin/darwin.nix else ../applications/networking/browsers/firefox-bin/linux.nix) {
```
Does the wrapper work on Darwin? Do we need it?

Atemu replied:
The darwin conditional should be for firefox-bin then.

firefox-bin-unwrapped is more of a plumbing thing. I don't know why it's exposed. It doesn't hurt I guess but it's not intended to be something a user would want to install I think.

Enzime replied:
Updated :+1:

Enzime committed:
firefox-{beta,devedition}-bin: add Mac sources

hraban commented:
I see some updates around unwrapped considered supported or not--What's the actual entrypoint of this PR now?

lafrenierejm commented:
> I see some updates around unwrapped considered supported or not--What's the actual entrypoint of this PR now?

The entrypoint is `firefox-bin` (_not_ the unwrapped variant).

lafrenierejm commented:
@Enzime Up for rebasing this?

Enzime commented:
I'm quite busy for the next 2 weeks, feel free to create a new PR, otherwise I'll rebase it when I get some more free time :+1:

Atemu commented on pkgs/top-level/all-packages.nix:
```diff
@@ -31595,40 +31595,56 @@ with pkgs;
-  firefox-bin = wrapFirefox firefox-bin-unwrapped {
-    pname = "firefox-bin";
-  };
+  firefox-bin = if stdenv.isDarwin then
+    callPackage ../applications/networking/browsers/firefox-bin/darwin.nix {
+      channel = "release";
+      generated = import ../applications/networking/browsers/firefox-bin/release_sources.nix;
+    }
+  else
+    wrapFirefox firefox-bin-unwrapped {
+      pname = "firefox-bin";
+    };
```
Hm, this kinda violates the DRY principle. Perhaps turning wrapFirefox into the identity function on Darwin would be simpler?

jameschensmith commented:
@Enzime, I'm so appreciative of your work on this PR (many thanks to the reviewers as well)! I haven't contributed to `nixpkgs` before, but I would love to help in any way get this merged.

One thought that I had was that the changes to `maintainers/scripts/update.py` could be isolated to a separate PR. As I'm writing this message, that file has not been touched in over a year, so we could possibly cherry-pick your commit from this branch into another branch and get an PR out there for that.

Enzime commented:
Closing this PR, anyone's who's interested feel free to pick this up

However note that `firefox-bin` is a lot less necessary as a source built version should be merged soon at #350384
